apache/superset
# feat: Live query validation in the SQL Lab UI

### CATEGORY

Choose one


### SUMMARY
This builds on #7422 to build check-as-you-type sql query validation in Sql Lab. This closes #6707 too.

This adds a (debounced) call to the `validate_sql_json` API endpoint with the querytext. If no validator is configured for the selected database, per the feature flag from #7422, no validation requests will be sent from the UI. On Lyft's infra the presto validator is able to return feedback to the user (end to end) in about 2 seconds.

At present feedback is provided only through the "annotations" mechanism build in to ACE, although I'd be open to adding full text elsewhere on the page if there's interest.

This diff does not include new cypress tests covering this interaction.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

![validation-med](https://user-images.githubusercontent.com/596042/57239173-60e18900-6fe8-11e9-99fe-086968204e70.gif)

Server error reporting:
<img width="921" alt="Screen Shot 2019-05-05 at 7 19 38 PM" src="https://user-images.githubusercontent.com/596042/57239184-663ed380-6fe8-11e9-93e0-c00883f3579b.png">

Query error reporting:
<img width="543" alt="Screen Shot 2019-05-05 at 7 21 08 PM" src="https://user-images.githubusercontent.com/596042/57239186-68089700-6fe8-11e9-8845-6dbbd6e3a810.png">

### TEST PLAN
1. Start a development instance
2. Follow the instructions in INSTALLING to enable the presto validator
3. Connect to a prestodb instance
4. Try writing some queries against it!

You can verify the unhappy paths by:

* Injecting 5xx into your communications with presto
* Using a database which does not support validation (like the built-in sqlite). Note that no validation attempts will even be made in this case.

### ADDITIONAL INFORMATION
- [X] Has associated issue: #6707

### REVIEWERS
@xtinec @DiggidyDave @khtruong @betodealmeida
